### PR TITLE
Build new API POC for OpenFisca

### PR DESCRIPTION
**Pull request by fpagnoux**
Connected to #100

### CODE REVIEW

**Comment by fpagnoux** [comment]
Feedback welcome on:

- @MattiSG: the test coverage.
- @l-vincent-l: the general architecture, as you're familiar with flask
- @cbenz: the code in general

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+sudo: false # Use new Travis container-based infrastructure
```
Comment (MattiSG) [inline .travis.yml]: Can't we use CircleCI for better performance?

File: README.md [hunk 2]
```diff
@@ -1 +1,23 @@
 # OpenFisca new Web-API
+
+## Install
+
+Will work once [this PR](https://github.com/openfisca/openfisca-core/pull/475) is merged. Until then, you need to install `openfisca-core` locally from the branch `variable-not-found`.
+
+```sh
+git@github.com:openfisca/openfisca-web-api.git
```
Comment (MattiSG) [inline README.md]: Missing command.

File: openfisca_web_api/app.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    parameters_description = {
+        name: { 'description' :parameter['description']}
+        for name, parameter in parameters.iteritems()
+        }
+
+    @app.route('/parameters')
+    def get_parameters():
+        return jsonify(parameters_description)
+
+
```
Comment (MattiSG) [inline openfisca_web_api/app.py]: Inconsistent whitespace (two blank lines).

File: tests/tests.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
+# -*- coding: utf-8 -*-
+
+from unittest import TestCase
+import json
+
+from openfisca_web_api.app import create_app
+
+tester = create_app('openfisca_dummy_country').test_client()
```
Comment (MattiSG) [inline tests/tests.py]: I like to call my [SUT](https://en.wikipedia.org/wiki/System_under_test) “subject”.

File: tests/tests.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+
+class ParametersRoute(TestCase):
+
+    @classmethod
+    def setUpClass(cls):
+        cls.response = tester.get('/parameters')
+
+    def test_return_code(self):
+        self.assertEqual(self.response.status_code, 200)
```
Comment (MattiSG) [inline tests/tests.py]: > Prefer a constant if possible, if Python provides something like `HTTP.OK` it's a good way to increase readability.

File: tests/tests.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+    def test_headers(self):
+        headers = self.response.headers
+        self.assertEqual(headers.get('Country-Package'), 'openfisca-dummy-country')
+        self.assertEqual(headers.get('Country-Package-Version'), '0.1.1')
```
Comment (MattiSG) [inline tests/tests.py]: Try to make test functions atomic:

```python
def test_package_name_header(self):
    self.assertEqual(self.response.headers.get('Country-Package'), 'openfisca-dummy-country'))

def test_package_version_header(self):
    self.assertEqual(self.response.headers.get('Country-Package-Version'), '0.1.1'))
```

…will yield much more informative failure messages and will scale more.

File: tests/tests.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    @classmethod
+    def setUpClass(cls):
+        cls.response = tester.get('/parameters')
+
+    def test_return_code(self):
+        self.assertEqual(self.response.status_code, 200)
+
+    def test_headers(self):
+        headers = self.response.headers
+        self.assertEqual(headers.get('Country-Package'), 'openfisca-dummy-country')
```
Comment (MattiSG) [inline tests/tests.py]: > The tests would be more resilient if the expected value was loaded dynamically from the Python object.

File: openfisca_web_api/app.py [hunk 8]
```diff
@@ -0,0 +1,40 @@
+# -*- coding: utf-8 -*-
+
+import os
+from flask import Flask, jsonify, abort
+
+from model import build_parameters, build_tax_benefit_system, build_headers
+
+def create_app(country_package = os.environ.get('COUNTRY_PACKAGE')):
+    if country_package is None:
+        raise ValueError(u"You must specify a country package to start the api. For instance, COUNTRY_PACKAGE=openfisca_france flask run")
```
Comment (cbenz) [inline openfisca_web_api/app.py]: Exception string must be encoded (in UTF-8), so either remove the `u` or add `.encode('utf-8')`

File: openfisca_web_api/app.py [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+
+    app = Flask(__name__)
+
+    tax_benefit_system = build_tax_benefit_system(country_package)
+    headers = build_headers(tax_benefit_system)
+    parameters = build_parameters(tax_benefit_system)
+    parameters_description = {
+        name: { 'description' :parameter['description']}
```
Comment (cbenz) [inline openfisca_web_api/app.py]: Fix spacing: `{'description': `

File: tests/tests.py [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+        self.assertEqual(headers.get('Country-Package-Version'), '0.1.1')
+
+    def test_item_content(self):
```
Comment (MattiSG) [inline tests/tests.py]: > Poor naming: what information can you take when CI tells you “failure in web API : test_item_content”? Prefer something like `test_response_data`.

File: openfisca_web_api/app.py [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+        for name, parameter in parameters.iteritems()
+        }
+
+    @app.route('/parameters')
+    def get_parameters():
+        return jsonify(parameters_description)
+
```
Comment (cbenz) [inline openfisca_web_api/app.py]: Keep only 1 empty line to be consistent

File: openfisca_web_api/model.py [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+    parsed_date = datetime.datetime.strptime(date, DATE_FORMAT)
+    next_day = parsed_date + datetime.timedelta(days=1)
+    return next_day.strftime(DATE_FORMAT)
+
+
+def build_parameter(parameter_json, parameter_path):
+    result = {}
+    result['description'] = parameter_json.get('description')
+    result['id'] = parameter_path
+    result['values'] = {}
```
Comment (cbenz) [inline openfisca_web_api/model.py]: Better initialize a dict with the 3 keys rather than doing 3 affectations.

File: openfisca_web_api/model.py [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+                parameters_json = parameters_json,
+                path_fragments = path_fragments + [child_name],
+                )
+
+
+def build_tax_benefit_system(country_package_name):
+    try:
+        country_package = importlib.import_module(country_package_name)
+    except ImportError:
+        raise ValueError(u"{} is not installed in your current environment".format(country_package_name))
```
Comment (cbenz) [inline openfisca_web_api/model.py]: Exception string must be encoded

File: setup.py [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+    description=u'POC of a new API for OpenFisca',
+    license='http://www.fsf.org/licensing/licenses/agpl-3.0.html',
+    url='https://github.com/openfisca/openfisca-web-api/tree/rewrite',
+    extras_require={
+        'test': [
+            'Openfisca-Dummy-Country >= 0.1.1',
+            'nose',
+            ],
+        },
+    include_package_data = True,  # Will read MANIFEST.in
```
Comment (cbenz) [inline setup.py]: Remove spaces around `=`

File: tests/tests.py [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+class ParameterRoute(TestCase):
+
+    def test_error_code_non_existing_parameter(self):
+        response = tester.get('/parameter/non.existing.parameter')
+        self.assertEqual(response.status_code, 404)
+
+    def test_return_code_existing_parameter(self):
+        response = tester.get('/parameter/impot.taux')
+        self.assertEqual(response.status_code, 200)
+
```
Comment (MattiSG) [inline tests/tests.py]: You might consider adding also a test on:

- `/parameter`
- `/parameters/`
- `/parameter/`
- `/parameter/with-ÜNı©ød€`
- `/parameter/with%20url%20encoding`
- `/parameter/impot.taux/`
- `/parameter/impot.taux/too-much-nesting`
- `/parameter//impot.taux/`

File: .travis.yml (hunk 1 above)
Comment (fpagnoux) [inline .travis.yml]: I thought we coud not have Circle in a branch and travis on another, but we should be able to get something playing with protected branches.

File: tests/tests.py (hunk 15 above)
Comment (fpagnoux) [inline tests/tests.py]: All these request return a `404`. What would these tests bring us ?

Comment (MattiSG) [inline tests/tests.py]: > All these requests return a `404`.

That's not the point. In software testing, you want to do [equivalence partitioning](https://en.wikipedia.org/wiki/Equivalence_partitioning) of the _input_ data, not the output.

> What would these tests bring us ?

Understanding of the limits, and future-proofing. The fact that they yield a 404 vs a 500 is informative, and this guarantee should be checked.

You could also enrich the dummy country package with the expected unicode field, if you want to support it.

File: Makefile [hunk 16]
```diff
@@ -0,0 +1,7 @@
+flake8:
+	@# Do not analyse .gitignored files.
+	@# `make` needs `$$` to output `$`. Ref: http://stackoverflow.com/questions/2382764.
```
Comment (MattiSG) [inline Makefile]: > Why do we need these `@`?

File: README.md [hunk 17]
```diff
@@ -1 +1,23 @@
 # OpenFisca new Web-API
+
+## Install
+
+Will work once [this PR](https://github.com/openfisca/openfisca-core/pull/475) is merged. Until then, you need to install `openfisca-core` locally from the branch `variable-not-found`.
```
Comment (MattiSG) [inline README.md]: Has been merged.

File: setup.py [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+
+from setuptools import setup, find_packages
+
+
+setup(
+    name='OpenFisca-Web-API-Rewrite',
+    version='0.1.0',
+    author='OpenFisca Team',
+    author_email='contact@openfisca.fr',
+    description=u'POC of a new API for OpenFisca',
```
Comment (MattiSG) [inline setup.py]: That kind of stuff stays forever. Don't say it's a PoC: this is a transient property that the version number should carry. State the goal: this is the OpenFisca API.

File: setup.py [hunk 19]
```diff
@@ -0,0 +1,30 @@
...
+    license='http://www.fsf.org/licensing/licenses/agpl-3.0.html',
```
Comment (MattiSG) [inline setup.py]: Redirects to https://www.gnu.org/licenses/agpl-3.0.html.

File: tests/tests.py [hunk 20]
```diff
@@ -0,0 +1,99 @@
+# -*- coding: utf-8 -*-
+
+from httplib import OK, NOT_FOUND
+import json
+import pkg_resources
+from nose.tools import assert_equal
+
+from openfisca_web_api.app import create_app
+
+TEST_COUNTRY_PACKAGE = 'openfisca_dummy_country'
```
Comment (MattiSG) [inline tests/tests.py]: > This is more accurately the `TEST_COUNTRY_PACKAGE_NAME`…

File: tests/tests.py [hunk 21]
```diff
@@ -0,0 +1,99 @@
...
+    parameters = json.loads(parameters_response.data)
+    assert_equal(
+        parameters[u'impot.taux'],
+        {u'description': u'taux d\'impôt sur les salaires'}
+        )
+
+
+def test_with_extra_slash():
+    response = subject.get('/parameters/')
+    assert_equal(response.status_code, NOT_FOUND)
```
Comment (MattiSG) [inline tests/tests.py]: > Is this really what we want?  :wink:
>
> Cheers for adding the test, no need to handle that in this PR if you're short on time.

File: tests/tests.py [hunk 22]
```diff
@@ -0,0 +1,99 @@
...
+        )
+
+
+def check_route_not_found(route):
+    response = subject.get(route)
+    assert_equal(response.status_code, NOT_FOUND)
+
+
+def test_wrong_routes():
+    wrong_routes = [
```
Comment (MattiSG) [inline tests/tests.py]: > They are not “wrong” as much as they are “not found”.
>
> I usually like to test those with an { associative array | map | object } that map the route to the expected status code. That makes for quite readable and scalable tests.

File: .travis.yml (hunk 1 above)
Comment (MattiSG) [inline .travis.yml]: Does this mean you don't intend to handle using Circle in this PR? It is indeed very simple to tell each provider which branches to build (would it be only by ticking “build only if `travis|circle.yml` is present”). I can help if you want.

File: Makefile (hunk 16 above)
Comment (cbenz) [inline Makefile]: In a Makefile, comments with `#` are displayed on `stdout`. `@` hides them.

File: openfisca_web_api/app.py [hunk 23]
```diff
@@ -0,0 +1,44 @@
...
+import os
+from flask import Flask, jsonify, abort
+
+from model import build_parameters, build_tax_benefit_system, build_headers
+
+
+def create_app(country_package = os.environ.get('COUNTRY_PACKAGE')):
+    if country_package is None:
+        raise ValueError(
+            u"You must specify a country package to start the api."
```
Comment (cbenz) [inline openfisca_web_api/app.py]: - Add a ` ` at the end of the first string, after the `.`
- api -> API

File: tests/tests.py (hunk 21 above)
Comment (cbenz) [inline tests/tests.py]: @fpagnoux [This stackoverflow answer](http://stackoverflow.com/a/33285603/3548266) explains you can set a global setting in flask: `app.url_map.strict_slashes = False` to accept trailing slashes.

File: README.md [hunk 24]
```diff
@@ -1 +1,21 @@
 # OpenFisca new Web-API
+
+## Install
+
+```sh
+git clone git@github.com:openfisca/openfisca-web-api.git
```
Comment (cbenz) [inline README.md]: Misses `cd openfisca-web-api`, sorry for not seeing it before ;-)

File: setup.cfg [hunk 25]
```diff
@@ -0,0 +1,5 @@
+[flake8]
+hang-closing = true
+max-line-length = 120
+; E251 unexpected spaces around keyword / parameter equals
```
Comment (cbenz) [inline setup.cfg]: Given we start from scratch, we can get rid of this ignored error, and follow PEP8 about the "no space around kwargs `=` operators` rule

File: setup.py [hunk 26]
```diff
@@ -0,0 +1,30 @@
...
+        'test': [
+            'Openfisca-Dummy-Country >= 0.1.1',
+            'nose',
+            'flake8',
+            ],
+        },
+    include_package_data=True,  # Will read MANIFEST.in
+    install_requires=[
+        'flask == 0.12',
+        'OpenFisca-Core >= 8.0.0, < 9.0',
```
Comment (cbenz) [inline setup.py]: Can you raise the compatibility to `10.0`?

File: README.md (hunk 24 above)
Comment (MattiSG) [inline README.md]: Always give `https` URLs. SSH-based ones are more prone to failing, as they depend on your user having a GitHub account.

File: tests/tests.py [hunk 27]
```diff
@@ -0,0 +1,96 @@
...
+def test_routes_robustness():
+    expected_codes = {
+        '/parameters/': OK,
+        '/parameter': NOT_FOUND,
+        '/parameter/': NOT_FOUND,
+        '/parameter/with-ÜNı©ød€': NOT_FOUND,
+        '/parameter/with%20url%20encoding': NOT_FOUND,
+        '/parameter/impot.taux/': OK,
+        '/parameter/impot.taux/too-much-nesting': NOT_FOUND,
+        '/parameter//impot.taux/': NOT_FOUND,
```
Comment (MattiSG) [inline tests/tests.py]: 👌

**Pull request merged**